kernel/src/snapshot.rs

```diff
@@ -67,6 +75,70 @@ impl Snapshot {
     }
 
+    /// Create a new [`Snapshot`] instance from an existing [`Snapshot`]. This is useful when you
+    /// already have a [`Snapshot`] lying around and want to do the minimal work to 'update' the
+    /// snapshot to a later version.
+    ///
+    /// # Parameters
+    ///
+    /// - `existing_snapshot`: reference to an existing [`Snapshot`]
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `version`: target version of the [`Snapshot`]. None will create a snapshot at the latest
+    ///   version of the table.
+    pub fn new_from(
+        existing_snapshot: Arc<Snapshot>,
+        engine: &dyn Engine,
+        version: Option<Version>,
+    ) -> DeltaResult<Arc<Self>> {
+        // simple heuristic for now:
+        // 1. if the new version < existing version, just return an entirely new snapshot
+        // 2. if the new version == existing version, just return the existing snapshot
+        // 3. list from existing snapshot version
+        // 4a. if new checkpoint is found: just create a new snapshot from that checkpoint (and
+        // commits after it)
+        // 4b. if no new checkpoint is found: do lightweight P+M replay on the latest commits
+        match version {
+            Some(v) if v < existing_snapshot.version() => {
+                Self::try_new(existing_snapshot.table_root().clone(), engine, version).map(Arc::new)
+            }
+            Some(v) if v == existing_snapshot.version() => Ok(existing_snapshot.clone()),
+            new_version => {
+                debug!(
+                    "new version: {new_version:?}, existing version: {}",
+                    existing_snapshot.version()
+                );
+                let log_root = existing_snapshot.log_segment.log_root.clone();
+                let fs_client = engine.get_file_system_client();
+
+                // create a log segment just from existing_snapshot.version -> new_version
+                let log_segment = LogSegment::for_versions(
+                    fs_client.as_ref(),
+                    log_root,
+                    existing_snapshot.version(),
+                    new_version,
+                )?;
+
+                if log_segment.has_checkpoint() {
+                    Self::try_new_from_log_segment(
+                        existing_snapshot.table_root().clone(),
+                        log_segment,
+                        engine,
+                    )
+                    .map(Arc::new)
+                } else {
+                    let (new_metadata, new_protocol) = log_segment.protocol_and_metadata(engine)?;
+                    let table_configuration = TableConfiguration::new_from(
+                        existing_snapshot.table_configuration(),
+                        new_metadata,
+                        new_protocol,
+                        log_segment.end_version,
+                    )?;
+                    Ok(Arc::new(Snapshot::new(log_segment, table_configuration)))
```

Comment:
I think we have to merge the existing log segment with the new slice, otherwise we would produce incorrect scans?

Reply:
yep was getting there :)